src/main/window_components/signalR.ts

```diff
@@ -10,48 +10,107 @@ let signalRConnection: HubConnection | null = null;
-            },
-        })
-        .configureLogging({
-            log(logLevel, message) {
-                signalrToElectronLog(log, logLevel, message);
-            },
-        })
-        .build();
-
-    signalRConnection.onclose((err) =>
-        log.error('SignalR connection has been lost', err)
-    );
-
-    signalRConnection.on('PendingAlerts', (newAlerts) => {
-        SetPendingAlerts({
-            alerts: newAlerts.map((a: any) => ({
-                id: a.id,
-                content: a.content,
-            })),
+): Promise<void> {
+    return new Promise((resolve, reject) => {
+        if (signalRConnection != null)
+            throw new Error('SignalR has already been initialized');
+
+        signalRConnection = new HubConnectionBuilder()
+            .withUrl(store.get('signalrUrl') as string, {
+                accessTokenFactory: () => store.get('apiKey') as string,
+            })
+            .withAutomaticReconnect({
+                nextRetryDelayInMilliseconds(retryContext) {
+                    log.warn('Retrying SignalR connection...');
+                    return Math.min(
+                        2_000 * retryContext.previousRetryCount,
+                        60_000
+                    );
+                },
+            })
+            .configureLogging({
+                log(logLevel, message) {
+                    signalrToElectronLog(log, logLevel, message);
+                },
+            })
+            .build();
+
+        signalRConnection.onclose((err) =>
+            log.error('SignalR connection has been lost', err)
+        );
+
+        signalRConnection.on('PendingAlerts', (newAlerts) => {
+            SetPendingAlerts({
+                alerts: newAlerts.map((a: any) => ({
+                    id: a.id,
+                    content: a.content,
+                })),
+            });
+            if (newAlerts.length > 0) createAlertsWindowCallback();
+        });
+
+        signalRConnection
+            .start()
+            .then(() => {
+                log.info('Connection to SignalR established');
+                return resolve();
+            })
+            .catch((err) => {
+                log.error('Error connecting to SignalR', err);
+                return reject(err);
+            });
+    });
+}
+
+/*
+* This function is used to invoke a SignalR event and wait for a response.
+* @param eventName The name of the event to invoke
+* @param eventResponse The name of the event to listen for the response
+* @param args Any arguments to pass to the event
+* @returns A promise that resolves with the response from the event
+*/
+function invokeExpectResponse<t>(eventName: string, eventResponse: string, ...args: any[]): Promise<t> {
+    return new Promise<t>((resolve, reject) => {
+        if (signalRConnection == null) {
+            reject();
+            return;
+        }
+
+        const listener = (response: t) => {
+            signalRConnection?.off(eventResponse, listener);
```

Comment:
Just a note with this: if an event comes through that doesn't have a listener registered, SignalR will start throwing warnings into the log. This function isn't being used right now so doesn't matter a ton